new: Add Python tier 2 and 3 support.

harlequin opened the pull request:
Support python as a tier level 3 language in moon.

- Added Python tier 3 support.
  - Will download and install Python into the toolchain when a `version` is configured.
  - Will parse the `requirements.txt` to resolve and install dependencies.
  - Added a `python.version` setting to `.moon/toolchain.yml`.
  - Added a `toolchain.python` setting to `moon.yml`.
  - Updated `moon bin` commands to support Python.

Would close Issues
- #1658 - per default an `virtual environment is used`
- #1595 - `poetry would be installed via requirements.txt`

milesj commented:
@harlequin Awesome stuff! I don't know Python at all, so I'll probably be asking a bunch of dumb questions.

Also, since this is a pretty large feature, it should merge into a dev branch. I've created `develop-1.30`, so can you update the merge base to that.

milesj commented on crates/config/src/toolchain/python_config.rs:
```diff
@@ -0,0 +1,36 @@
+    /// Location of the WASM plugin to use for Python support.
+    pub plugin: Option<PluginLocator>,
+
+    /// Options for pnpm, when used as a package manager.
```
Says pnpm.

milesj commented on crates/config/src/toolchain/python_config.rs:
```diff
@@ -0,0 +1,36 @@
+    /// The relative root of the virtual environment workspace. Default to moon's
+    /// workspace root
+    #[setting(default = ".", skip)]
+    pub venv_root: String,
```
How do venv's work in python? Are there multiple or just one? Or per project? etc.

harlequin replied:
virtual environments can be created per project

But I like the idea to have the single version policy, maybe I am just removing this setting.

milesj commented on legacy/python/lang/src/pip_requirements.rs:
```diff
@@ -0,0 +1,15 @@
+pub fn load_lockfile_dependencies(path: PathBuf) -> miette::Result<LockfileDependencyVersions> {
+    let mut deps: LockfileDependencyVersions = FxHashMap::default();
+
+    let lockfile = fs::read_to_string(path.as_path()).expect("Unable to read file");
```
You'll need an actual `requirements.txt` parser and the https://crates.io/crates/requirements crate is too old.

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,129 @@
+            let args = vec![
+                "-m",
+                "venv",
+                virtual_environment.as_os_str().to_str().unwrap(),
```
Does this need an absolute path? Could it just be `&python.config.venv_name`?

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,129 @@
+            );
+        }
+
+        if let Some(pip_version) = &pip_config.version {
```
This shouldn't be done here. Use the setup toolchain action instead.

harlequin replied:
Is this not the same like bins in the rust toolchain?
Because also here it will be executed in the install_deps action.

harlequin replied:
I have had a look into this topic. From what I understand is that the function setup_toolchain is used for installing python via proto onto the system (~/.proto).

I have implemented now the virtual environment into the setup_tools function, because this will also be called when doing a moon sync command.

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,129 @@
+
+    // python.exec_python(args, working_dir)
+
+    if let Some(pip_config) = &python.config.pip {
```
This may make more sense in the setup toolchain action.

harlequin replied:
Some as above

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,129 @@
+            );
+        }
+
+        let requirements_path = find_requirements_txt(working_dir, &get_workspace_root());
```
Always use `working_dir`.

milesj replied:
Need to remove all these `get_workspace_root` calls everywhere.

harlequin replied:
Done

milesj commented on legacy/python/tool/src/python_tool.rs:
```diff
@@ -0,0 +1,279 @@
+            //TODO: only read from root, but ready for sub virtual environments
+            if let Some(requirements_path) = fs::find_upwards_until(
+                "requirements.txt",
+                get_workspace_root(),
```
`working_dir, &self.workspace_root`

harlequin replied:
Done

harlequin commented on crates/config/src/toolchain/python_config.rs:
```diff
@@ -0,0 +1,36 @@
+    pub install_args: Option<Vec<String>>,
+
+    /// The version of pip to download, install, and run `pip` tasks with.    
+    pub version: Option<UnresolvedVersionSpec>,
```
pip is a library from python, which is shipped automatically. But pip itself can be upgraded independently of the python version. So the scenario would be, for example, take python 3.11 (which shipps an older pip version) and upgrade pip afterwards.

So I would like to keep this.

milesj commented:
Can you rebase `develop-1.30` so that the diff is nice.

harlequin commented:
Hey @milesj 

I have done the re-work you have suggested and from my point of view it shall be complete.
Therefore I have removed the WIP tag on this commit.

milesj commented on .yarn/versions/3fbef0f1.yml:
```diff
@@ -1,4 +1,9 @@
 releases:
+  "@moonrepo/nx-compat": minor
```
I don't release these unless required, you can mark them as declined like this: https://github.com/moonrepo/moon/blob/master/.yarn/declined.txt

harlequin replied:
Just wanted that the pipeline will be executed successfully, I have deleted this file and let the release versioning on your side.

harlequin replied:
Rebased to current develop-1.30

milesj commented on CHANGELOG.md:
```diff
@@ -28,6 +65,15 @@
 
 - Fixed an issue running `npx` commands (primarily through lockfile deduping).
 
+## Unreleased
```
Bad merge

harlequin replied:
Done

milesj commented on legacy/python/lang/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+moon_lang = { path = "../../core/lang" }
+moon_logger = { path = "../../core/logger" }
+cached = { workspace = true }
+cargo-lock = "9.0.0"
```
These deps need to be cleaned up

harlequin replied:
Done

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,76 @@
+            if install_args.iter().any(|x| !x.starts_with("-")) && requirements_path.is_none() {
+                console.out.print_checkpoint(
+                    Checkpoint::Setup,
+                    "Install pip dependencies from install args",
```
Change this to just "pip install" to match the other tools

harlequin replied:
Done

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,76 @@
+                )?;
+
+                let mut args = vec!["-m", "pip", "install"];
+                if pip_config.install_args.is_some() {
```
Do `if let Some(args) = &pip_config.install_args {` for these kind of patterns.

harlequin replied:
Done

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,76 @@
+        if let Some(req) = requirements_path {
+            console.out.print_checkpoint(
+                Checkpoint::Setup,
+                format!("Install pip dependencies from {}", color::path(&req)),
```
"pip install" here too

harlequin replied:
Done

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,76 @@
+            )?;
+
+            let mut args = vec!["-m", "pip", "install"];
+            if pip_config.install_args.is_some() {
```
Same if-let here

harlequin replied:
Done

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,76 @@
+            args.extend(["-r", req.as_os_str().to_str().unwrap()]);
+
+            operations.push(
+                Operation::task_execution(format!(" {}", args.join(" ")))
```
"python <args>"

harlequin replied:
Done

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,76 @@
+                }
+
+                operations.push(
+                    Operation::task_execution(format!(" {}", args.join(" ")))
```
"python <args>"

harlequin replied:
Done

milesj commented on legacy/python/platform/src/actions/setup_tool.rs:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    // Create version file
+    if let Some(python_version) = &python.config.version {
```
This should be behind a setting, like `syncVersionManagerConfig` (like node does)

harlequin replied:
There is no external version manager planned, therefore this is removed.

milesj commented on legacy/python/platform/src/actions/setup_tool.rs:
```diff
@@ -0,0 +1,50 @@
+    if let Some(pip_config) = &python.config.pip {
+        let virtual_environment = &workspace_root.join(python.config.venv_name.clone());
+
+        if !virtual_environment.exists() {
```
Is venv coupled with a requirements.txt? Or are they separate?

If venv stuff can be per-project, this venv logic should be in the sync_project action, since the setup_tool doesn't run in the project directory.

harlequin replied:
As mentioned below, strategy changed and now working with single version policy and project specific configurations.

milesj commented on legacy/python/platform/src/manifest_hash.rs:
```diff
@@ -0,0 +1,10 @@
+
+hash_content!(
+    pub struct PythonManifestHash {
+        // pub dependencies: BTreeMap<String, DependencyDetail>,
```
Still needs to be implemented.

harlequin replied:
Was dead code

harlequin commented:
I have updated today again main parts of the implementation, and also re-thought about the setup_tools and other areas.
I have now re-worked that the venv and deps install will be executed on the install_deps phase which is absolutely sufficient.

Additional I have removed all the dead code and not needed comments.

These are the areas were I have had a look into ... hope that I didn't missed something ;)

- remove yarn version modification
- remove version configuration for pip, if user want's to update the pip dependency he can do this via the requirements.txt or installArgs method
- add root_requirements_only, to be able to create venv in project scope or in workspace scope
- pip_requirements fixed the occupied/vaccant
- clean up Cargo.toml
- reworked and rethinked the install_deps and setup_tool function; install_deps phase is absolute sufficient
- remove not needed comments
- remove all references to get_workspace_root
- Cleaned up CHANGELOG.md

harlequin commented on legacy/python/tool/src/python_tool.rs:
```diff
@@ -0,0 +1,279 @@
+        project_root: &Path,
+    ) -> miette::Result<LockfileDependencyVersions> {
+        let Some(lockfile_path) =
+            fs::find_upwards_until("requirements.txt", project_root, get_workspace_root())
```
Done

milesj commented on crates/config/src/toolchain/python_config.rs:
```diff
@@ -0,0 +1,33 @@
+    #[setting(nested)]
+    pub pip: Option<PipConfig>,
+
+    #[setting(default = ".venv", skip)]
```
Can you add a description then run `just schemas`.

harlequin replied:
Will do

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,80 @@
+            );
+        }
+
+        if let Some(install_args) = &pip_config.install_args {
```
This should probably still `pip install` even if there are no custom args yeah? The args from the config should just append.

harlequin replied:
Yes you are right, even with the next comment (difference first and second pip install)
Therefore I have re-worked the strategy that in before the `args` will be constructed.

And if these are not empty, only one `pip install` will be executed.

milesj commented on legacy/python/platform/src/actions/install_deps.rs:
```diff
@@ -0,0 +1,80 @@
+            }
+        }
+
+        if let Some(req) = requirements_path {
```
Actually what's the difference between this block, and the one above? They both pip install.

harlequin replied:
See above

milesj commented on legacy/python/platform/src/python_platform.rs:
```diff
@@ -0,0 +1,311 @@
+        hasher: &mut ContentHasher,
+        _hasher_config: &HasherConfig,
+    ) -> miette::Result<()> {
+        if let Some(python_version) = &self.config.version {
```
We should probably still hast he dependencies, even if no version is configured.

harlequin replied:
Okay

milesj commented on legacy/python/platform/src/python_platform.rs:
```diff
@@ -0,0 +1,311 @@
+        hasher: &mut ContentHasher,
+        _hasher_config: &HasherConfig,
+    ) -> miette::Result<()> {
+        if let Some(python_version) = &self.config.version {
```
Same here.

harlequin replied:
Okay

milesj commented:
@harlequin Awesome work, just left a few more comments/questions.

milesj commented:
@harlequin Looks good! Just need to fix the conflict.

The docker check can be ignored. I don't think that works in forks.

milesj commented:
Awesome stuff, thanks for all the hard work!!!
